app/scripts/lib/sentry.js

```diff
@@ -0,0 +1,133 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+define([
+  'underscore',
+  'raven',
+  'lib/url'
+], function (_, Raven, Url) {
+
+  var ALLOWED_QUERY_PARAMETERS = [
+    'client_id',
+    'keys',
+    'redirect_uri',
+    'scope',
+    'style',
+    'verification_redirect',
+    'campaign',
+    'customizeSync',
+    'entrypoint',
+    'context',
+    'migration',
+    'service',
+    'setting',
+    'automatedBrowser'
+  ];
+
+  /**
+   * function that gets called before data gets sent to error metrics
+   *
+   * @param {Object} data
+   *  Error object data
+   * @returns {Object} data
+   *  Modified error object data
+   * @private
+   */
+  function beforeSend(data) {
+    if (data && data.request && data.request.url) {
+      data.request.url = cleanUpQueryParam(data.request.url);
+    }
+
+    return data;
+  }
+
+  /**
+   * Sanitize and clear extra query parameter information
+   *
+   * @param {String} url
+   * @returns {String} url
+   * @private
+   */
+  function cleanUpQueryParam(url) {
+    var startOfParams = url.indexOf('?');
+    var newUrl = url;
+    var params;
+
+    if (startOfParams >= 0) {
+      params = Url.searchParams(url.substring(startOfParams + 1));
+      newUrl = url.substring(0, startOfParams);
+    }
+
+    if (_.isObject(params)) {
+      Object.keys(params).forEach(function (key) {
+        // if the param is a PII (not allowed) then reset the value.
+        if (! _.contains(ALLOWED_QUERY_PARAMETERS, key)) {
+          params[key] = 'VALUE';
+        }
+      });
+
+      newUrl += Url.objToSearchString(params);
+    }
+
+    return newUrl;
+  }
+
+  /**
+   * Creates a SentryMetrics object that starts up Raven.js
+   *
+   * Read more at https://github.com/getsentry/raven-js
+   *
+   * @param {String} host
+   * @constructor
+   */
+  function SentryMetrics (host) {
+    if (host) {
+      this._endpoint = '//__API_KEY__@' + host + '/metrics-errors';
```

Comment:
> At one point pushing __API_KEY__ into the server was discussed. Is that unfeasible?

Yes, only the server has access to the api server. 

> I see it's replaced down here, but I don't follow how it works. Here the __API_KEY__ is in the endpoint, down there __API_KEY__ is replaced in the query parameters.

This is because Raven.js expects the real api key to be on the client. for enhanced security we mask it here with __API_KEY__ and because we have our own endpoint then we can replace it on the server. Other developers do not do this, they send Sentry reports directly. In our case to avoid external requests we use our own endpoint. I can either add a more descriptive comment OR update our version of raven.js and force it to exclude the API_KEY in the client library.

@shane-tomlinson which do you prefer?